salt/_modules/metalk8s.py

```diff
@@ -355,3 +358,164 @@ def cmp_sorted(*args, **kwargs):
         kwargs['key'] = functools.cmp_to_key(kwargs.pop('cmp'))
 
     return sorted(*args, **kwargs)
+
+
+def _error(ret, err_msg):
+    ret["result"] = False
+    ret["comment"] = err_msg
+    return ret
+
+
+def _atomic_copy(source, dest, tmp_prefix="."):  # pragma: no cover
+    """Minimalistic implementation of an atomic copy operation.
+
+    First, we write the contents of the source file to a temporary file, which
+    is saved in the same directory as the target, with a custom prefix.
+    Finally, we link the temporary file contents to the destination filename.
+    """
+    base_name = os.path.basename(dest)
+    dir_name = os.path.dirname(dest)
+
+    with open(source, mode='rb') as src_file:
+        contents = src_file.read()
+
+    with tempfile.NamedTemporaryFile(
+        prefix=tmp_prefix, dir=dir_name, delete=False,
+    ) as tmp_file:
+        tmp_file.write(contents)
```

Comment:
`manage_static_pod_manifest` sets some permissions on the final file, after the file has been written. However, this means we potentially/likely have a file with different permissions at some point in time, in which target data is written.

As such, I'd suggest to set the perms etc. on `tmp_file` before writing any data into it.

Reply:
Right, good point, so I guess we only need the `file.check_perms` call if the source and target checksums match.

Reply:
From the comments in the state implementation, I have the impression `ret` is passed to `check_perms` which is idempotent, so that can stay as-is. I'd set the file perms/owner to whatever is later passed to `file.check_perms` later on, at this point.

Reply:
However, it means I need to compare the desired perms/owner with the existing file if there is one (otherwise `check_perms` would not add anything to `ret["changes"]`).

Reply:
How so? `check_perms` sets the uid/gid/attrs to the desired ones, unless they're already set to the desired ones (in which cases there are no 'changes' reported). This is unconditionally called in `manage_static_pod_manifest` as using root/root/0600. Now, in this function *we are always creating a new file*, so there *are* 'changes' (new file created). We don't really need/want `check_perms` to 'add' more changes since we can as well create this new file using the attributes we want it to have anyway.

So, the 'changes' of the `manage_static_pod_manifest` call are either 'file created/replaced (using the content and attrs we want it to have)', or 'file not modified, but existing file attrs changed to desired values because someone changed them to something else for whatever reason', or 'no changes (file contents and attrs are what they should be)'.